### PR TITLE
[DT-1253] Check for empty and null arguments for `findVoteUsersByElectionReferenceIdList`

### PR DESCRIPTION
**Pull request by fboulnois**
### Addresses

https://broadworkbench.atlassian.net/browse/DT-1253

### Summary

Add tests and fix for when `findVoteUsersByElectionReferenceIdList` is empty or null

----
Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java [hunk 1]
```diff
@@ -657,13 +658,21 @@ public DarCollection cancelDarCollectionElectionsAsChair(DarCollection collectio
...
-  public DarCollection createElectionsForDarCollection(User user, DarCollection collection) {
+  public DarCollection createElectionsForDarCollection(User user, DarCollection collection)
+      throws Exception {
     try {
       List<String> createdElectionReferenceIds = collectionServiceDAO.createElectionsForDarCollection(
           user, collection);
-      List<User> voteUsers = voteDAO.findVoteUsersByElectionReferenceIdList(
-          createdElectionReferenceIds);
+      if (createdElectionReferenceIds == null || createdElectionReferenceIds.isEmpty()) {
+        var e = new IllegalStateException(
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java:667]: Nice

File: src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java [hunk 2]
```diff
@@ -657,13 +658,21 @@ public DarCollection cancelDarCollectionElectionsAsChair(DarCollection collectio
...
    */
-  public DarCollection createElectionsForDarCollection(User user, DarCollection collection) {
+  public DarCollection createElectionsForDarCollection(User user, DarCollection collection)
+      throws Exception {
     try {
       List<String> createdElectionReferenceIds = collectionServiceDAO.createElectionsForDarCollection(
           user, collection);
-      List<User> voteUsers = voteDAO.findVoteUsersByElectionReferenceIdList(
-          createdElectionReferenceIds);
+      if (createdElectionReferenceIds == null || createdElectionReferenceIds.isEmpty()) {
```
Comment (pshapiro4broad) [inline src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java]: How can the result of `createElectionsForDarCollection()` ever be `null`? When calling an `internal` API that returns a collection type, we should be able to assert that we will never see `null`, only an empty collection.

File: src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java [hunk 3]
```diff
@@ -657,13 +658,21 @@ public DarCollection cancelDarCollectionElectionsAsChair(DarCollection collectio
...
+        var e = new IllegalStateException(
+            "No elections were created for DAR Collection: %s".formatted(
+                collection.getDarCode()));
+        logException(e);
+        throw e;
+      }
       try {
```
Comment (pshapiro4broad) [inline src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java:673]: Nesting `try`s should be avoided when possible. In this case, both the inner and outer `try` catch `Exception`, so the code should work the same if this nested `try` of moved after the outer `catch` to line 677 below

File: src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java [hunk 4]
```diff
@@ -673,6 +682,7 @@ public DarCollection createElectionsForDarCollection(User user, DarCollection co
     } catch (Exception e) {
       logException("Exception creating elections and votes for collection: %s".formatted(
           collection.getDarCollectionId()), e);
+      throw e;
```
Comment (pshapiro4broad) [inline src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java:685]: This is a change in API behavior, is that part of this ticket? Was something broken by not throwing an exception which is now fixed?

File: src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java [hunk 5]
```diff
@@ -657,13 +658,21 @@ public DarCollection cancelDarCollectionElectionsAsChair(DarCollection collectio
...
       List<String> createdElectionReferenceIds = collectionServiceDAO.createElectionsForDarCollection(
           user, collection);
-      List<User> voteUsers = voteDAO.findVoteUsersByElectionReferenceIdList(
-          createdElectionReferenceIds);
+      if (createdElectionReferenceIds == null || createdElectionReferenceIds.isEmpty()) {
+        var e = new IllegalStateException(
+            "No elections were created for DAR Collection: %s".formatted(
+                collection.getDarCode()));
+        logException(e);
+        throw e;
```
Comment (pshapiro4broad) [inline src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java:671]: I've mentioned this before and I don't know how common it is in the duos codebase, but in general if you `throw`, you shouldn't `log`, since any `catch` which doesn't `throw` should be logging already. Typically, logging before `throw` causes the exception to be logged twice.

**Review by rushtong — APPROVED** [review]
Looks reasonable to me, thank you 👍🏽

File: src/test/java/org/broadinstitute/consent/http/service/DarCollectionServiceTest.java [hunk 6]
```diff
@@ -312,6 +308,55 @@ void testCreateElectionsForDarCollection() throws Exception {
...
+
+  @Test
+  void testCreateElectionsForDarCollectionNull() throws Exception {
+    User user = new User();
+    user.setEmail("email");
+    DataAccessRequest dar = new DataAccessRequest();
+    dar.setReferenceId(UUID.randomUUID().toString());
+    DarCollection collection = createMockCollections(1).get(0);
+    collection.setDars(Map.of(dar.getReferenceId(), dar));
+    when(darCollectionServiceDAO.createElectionsForDarCollection(any(), any())).thenReturn(null);
```
Comment (pshapiro4broad) [inline src/test/java/org/broadinstitute/consent/http/service/DarCollectionServiceTest.java]: I believe this is the same as not mocking at all, since all mocked APIs return `null` by default. If you only want to verify that an API is called (which mocking will do if you have strict stubbing enabled), I would instead use `verify` after the API is called, to make the intent of the code clearer.

File: src/test/java/org/broadinstitute/consent/http/service/DarCollectionServiceTest.java [hunk 7]
```diff
@@ -312,6 +308,55 @@ void testCreateElectionsForDarCollection() throws Exception {
...
+    dar.setReferenceId(UUID.randomUUID().toString());
+    DarCollection collection = createMockCollections(1).get(0);
+    collection.setDars(Map.of(dar.getReferenceId(), dar));
+    when(darCollectionServiceDAO.createElectionsForDarCollection(any(), any())).thenReturn(
+        List.of("electionId"));
+    when(voteDAO.findVoteUsersByElectionReferenceIdList(any())).thenThrow(
+        IllegalArgumentException.class);
+
+    service.createElectionsForDarCollection(user, collection);
+    verify(darCollectionServiceDAO, times(1)).createElectionsForDarCollection(any(), any());
```
Comment (pshapiro4broad) [inline src/test/java/org/broadinstitute/consent/http/service/DarCollectionServiceTest.java]: `times(1)` is the default so you can omit it
```suggestion
    verify(darCollectionServiceDAO).createElectionsForDarCollection(any(), any());
```

Also I would encourage matching against more specific arguments than `any()` if possible. You can use `ArgumentCaptor` to assert on a subfield of an argument in cases where you can't match the entire argument.

File: src/test/java/org/broadinstitute/consent/http/service/DarCollectionServiceTest.java [hunk 8]
```diff
@@ -312,6 +308,55 @@ void testCreateElectionsForDarCollection() throws Exception {
...
+    verify(voteDAO, times(1)).findVoteUsersByElectionReferenceIdList(any());
+    verify(emailService, times(0)).sendDarNewCollectionElectionMessage(any(), any());
```
Comment (pshapiro4broad) [inline src/test/java/org/broadinstitute/consent/http/service/DarCollectionServiceTest.java]: If `emailService` isn't used at all, I would use `verifyNoInteractions()` instead
```suggestion
    verifyNoInteractions(emailService);
```

File: src/test/java/org/broadinstitute/consent/http/db/VoteDAOTest.java [hunk 9]
```diff
@@ -411,13 +412,27 @@ void testUpdateRationaleByVoteIds() {
...
+  void testFindVoteUsersByElectionReferenceIdList_Empty() {
+    assertThrows(IllegalArgumentException.class, () -> {
+      voteDAO.findVoteUsersByElectionReferenceIdList(List.of());
+    });
+  }
+
+  @Test
+  void testFindVoteUsersByElectionReferenceIdList_Null() {
+    assertThrows(IllegalArgumentException.class, () -> {
+      voteDAO.findVoteUsersByElectionReferenceIdList(null);
```
Comment (pshapiro4broad) [inline src/test/java/org/broadinstitute/consent/http/db/VoteDAOTest.java:432]: I suppose there's no harm in this, but why test a case which can't arise in the code base? I think it's better to only test paths which a user can encounter.

**Review by pshapiro4broad — APPROVED** [review]
I have a few comments. I think the fix is OK but I don't think it's helpful to code for cases which can't arise out of database state or user interactions.

File: src/test/java/org/broadinstitute/consent/http/service/DarCollectionServiceTest.java (hunk 6 above)
Comment (fboulnois) [reply src/test/java/org/broadinstitute/consent/http/service/DarCollectionServiceTest.java]: It seems to return an empty list by default, hence the separate test for `null`.

File: src/test/java/org/broadinstitute/consent/http/db/VoteDAOTest.java (hunk 9 above)
Comment (fboulnois) [reply src/test/java/org/broadinstitute/consent/http/db/VoteDAOTest.java:432]: This helped identify the problem in the first place: if the arguments to `findVoteUsersByElectionReferenceIdList` are `null` or empty, then this will throw the error that we saw from Sentry. This also helped me think through a solution to the issue.

In the future, it is possible that we refactor `createElectionsForDarCollection` and accidentally return a `null`, which would cause the subsequent call to `findVoteUsersByElectionReferenceIdList` to fail. By having explicitly added this test, I think we signal to future developers (e.g. myself) that this state cannot be valid and/or otherwise they need to also rethink how `findVoteUsersByElectionReferenceIdList` works.

In general, I like the idea of exhaustively testing the valid and invalid states of a method, even if they are not currently reachable. This encodes the assumptions that we are truly making about the code, and if those assumptions change and we refactor the code, then we also need to update the tests that broke as a result.

For these reasons I think it's worthwhile to keep this test.

File: src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java (hunk 4 above)
Comment (fboulnois) [reply src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java:685]: Yes, the error wasn't being percolated up to the caller and it continued on and called `darCollectionDAO.findDARCollectionByCollectionId`, which should not occur in this case.

File: src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java (hunk 3 above)
Comment (fboulnois) [reply src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java:673]: @rushtong suggested I scope `findVoteUsersByElectionReferenceIdList` to ignore email failures but not election creation errors. I otherwise didn't modify the logic here.

File: src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java (hunk 5 above)
Comment (fboulnois) [reply src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java:671]: If I comment out the `throw`s, this causes the `null` and `empty` tests to fail, which are the states we want to avoid.

File: src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java (hunk 3 above)
Comment (pshapiro4broad) [reply src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java:673]: Yes, my point is that the original code was nesting `try`s where it didn't need to do so.

File: src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java (hunk 5 above)
Comment (pshapiro4broad) [reply src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java:671]: There's no need to remove the `throw` if that's the behavior we want, my suggestion was to either `throw` or `log` but not both, as that will likely result in the exception being logged twice. But that's assuming that the consent service is logging in the root exception handler, which may not be the case.

Looking at `Resource.createExceptionResponse()`, unknown exceptions are logged, so it appears that double logging will occur. This code throws `IllegalStateException` but I don't see a custom handler for it in `Resource`, so it should end up handled by the `Exception.class` case.

File: src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java (hunk 4 above)
Comment (pshapiro4broad) [reply src/main/java/org/broadinstitute/consent/http/service/DarCollectionService.java:685]: OK, I think see. What I'd expect to see here is a catch that rethrows as a more specific exception and wrap the previous one as a cause. The log message would instead be the new exception's message. And finally if we wanted to log, we could do it in the top level catch handler.

But it looks like that's not how the consent service works right now, so I agree it's best to keep with the current process.

I do think nesting `try`s isn't great though, and now that this method `throws Exception` you no longer need the outer `catch`

File: src/test/java/org/broadinstitute/consent/http/service/DarCollectionServiceTest.java (hunk 6 above)
Comment (pshapiro4broad) [reply src/test/java/org/broadinstitute/consent/http/service/DarCollectionServiceTest.java]: You're right, I didn't know that about mockito. That makes sense though since I don't think collection APIs should ever return nulls in the first place..
